animatplot/blocks/lineplots.py

```diff
@@ -21,41 +26,100 @@ class Line(Block):
+            y = args[0]
+            x = None
+        elif len(args) == 2:
+            [x, y] = args
+        else:
+            raise ValueError("Invalid data arguments to Line block")
+
+        if y is None:
+            raise ValueError("Must supply y data to plot")
+        y = np.asanyarray(y)
+        if str(y.dtype) == 'object':
+            # ragged array
+            if x is None:
+                raise ValueError("Must specify x data explicitly when passing"
+                                 "a ragged array for y data")
+            else:
+                x = np.asanyarray(x)
 
-    def _update(self, i):
-        Slice = self._make_slice(i, 2)
-        x_vector = self.x[Slice]
-        y_vector = self.y[Slice]
+            if not all(len(xline) == len(yline) for xline, yline in zip(x, y)):
+                raise ValueError("Length of x & y data must match one another "
+                                 "for every frame")
+            else:
+                self._is_list = True
+        else:
+            # Rectangular data
+            if y.ndim != 2:
+                raise ValueError("y data must be 2-dimensional")
+
+            # x is optional
+            shape = list(y.shape)
+            shape.remove(y.shape[t_axis])
+            data_length, = shape
+            if x is None:
+                x = np.arange(data_length)
+            else:
+                x = np.asanyarray(x)
+
+            shape_mismatch = "The dimensions of x must be compatible with " \
+                             "those of y, but the shape of x is {} and the " \
+                             "shape of y is {}".format(x.shape, y.shape)
+            if x.ndim == 1:
+                # x is constant over time
+                if len(x) == data_length:
+                    # Broadcast x to match y
+                    x = np.expand_dims(x, axis=t_axis)
+                    x = np.repeat(x, repeats=y.shape[t_axis], axis=t_axis)
+                else:
+                    raise ValueError(shape_mismatch)
+            elif x.ndim == 2:
+                if x.shape != y.shape:
+                    raise ValueError(shape_mismatch)
+            else:
+                raise ValueError("x, must be either 1- or 2-dimensional")
+
+        self.x = x
+        self.y = y
+
+        frame_slice = self._make_slice(i=0, dim=2)
+
+        x_first_frame_data = self.x[frame_slice]
+        y_first_frame_data = self.y[frame_slice]
 
+        self.line, = self.ax.plot(x_first_frame_data,
+                                  y_first_frame_data, **kwargs)
+
+    def _update(self, frame):
+        frame_slice = self._make_slice(frame, dim=2)
+        x_vector = self.x[frame_slice]
+        y_vector = self.y[frame_slice]
         self.line.set_data(x_vector, y_vector)
-        return self.line
 
     def __len__(self):
-        if self._is_list:
-            return self.x.shape[0]
-        return self.x.shape[self.t_axis]
+        return self.y.shape[self.t_axis]
```

Comment:
Without the ```if self._is_list:``` if a user passes in a value for ```t_axis```, and a ragged array, this method will error. The docstring explicitly says that if the user passes in a list of numpy arrays, then t_axis will be ignored. I suppose that one option would be to set ```self.t_axis=0``` in the check for ```if str(y.dtype) == object``` which might be cleaner.